backend/logic/users/users.js

```diff
@@ -167,10 +169,76 @@ var getCompletedCourses = async studentID => {
   return courses;
 };
 
+var login = async () => {
+	
+	let LocalStrategy = require('passport-local').Strategy;
+		
+	//LOCAL SIGNIN
+	app.passport.use('local-signin', new LocalStrategy(  
+	{
+		// by default, local strategy uses username and password, we will override with email
+		usernameField : 'username',
+		passwordField : 'password',
+		passReqToCallback : true // allows us to pass back the entire request to the callback
+	},
+	
+	function(req, username, password, done) {
+		let error;
+
+		let isValidPassword = function(userpass, password){
+			return hasher.hashPass(password) === userpass; 
+		}
+		// Check for missing input params
+		if (!username) {
+			error = 'Undefined username';
+		} else if (!password) {
+			error = 'Undefined password';
+		}
+
+		if (error) {
+			return done(error);
+		}
+
+		// Check for invalid formatting
+		if (!format.verifyUsername(username)) {
+			error = 'Invalid format username';
+		} else if (!format.verifyPassword(password)) {
+			error = 'Invalid format password';
+		}
+
+		if (error) {
+			return done(error);
+		}
```

Comment:
Nit: on line 204 and 206 you can just write return done(the messages you have written)